### ISSUE TITLE
GitHub Actions: Fixes to adding-labels-to-pull-request action

### ISSUE THREAD

**Issue by Aveline-art** [hackforla/website#1787]
### Overview

The recent GitHub action (adding labels to pull request) is currently lacking a couple of key features and fixes. Implement the features listed in the below action items.

### Action Items
- [x] Look for an alternate to the pull_request_target event to prevent malicious code injections into GitHub action execution
- [x] ~~Have the code exit successfully, but quietly in the event of no linked issue~~ Note: abandoned due to exiting loudly being more preferable.
- [x] Investigate why this action does not run on pull request merging forks to non-default branches (aka wins-feature-1)
  - [x] Then, fix the action such that it runs on all branches
- [ ] Awaiting confirmation: filter out status labels from linked issue so they do not appear on the pull request
- [x] ~~Use the needs: workflow syntax for better control flow~~ **Note:** No longer applicable as there is only 1 job. Using multiple artifacts to transfer info between jobs will only muddle the code.
- [ ] Test whether the repository secret is required for workflow to function in the H4LA repo
  - [ ] In the event that this testing requires directly editing the yml file, please ask a senior coder for permission to do so! All tests needs a date and description a comment below. This will leave a trail for us in case something breaks!
- [x] Ensure code maintainability and readability.
  - [x] Variables names need to be intuitive and consistent
  - [x] Leave comments when needed
  - [x] Leave notes when needed
- [x] Double check for [injection attacks concerns](https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#understanding-the-risk-of-script-injections)

### Checks
- [x] Check result with one linked issue: return success, labels added
- [x] Check result with no linked issue: return error, no labels
- [x] Check result with more than 1 linked issue: return error, no labels
- [x] Check result when the number of linked issues changes: 
  - [x] 0 to 1: return success, labels added
  - [x] 0 to 2+: return error, no labels
  - [x] 1 to 0: return error, labels stays
  - [x] 1 to 2+: return error, labels stays
  - [x] 2+ to 0: return error, no labels
  - [x] 2+ to 1: return success, labels added

### Resources/Instructions
[Keeping your GitHub Actions and workflows secure: Preventing pwn requests](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)
[Events that trigger workflows](https://docs.github.com/en/actions/reference/events-that-trigger-workflows)
[octokit/rest.js](https://octokit.github.io/rest.js/v18)
[Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions)
[Storing workflow data as artifacts](https://docs.github.com/en/actions/guides/storing-workflow-data-as-artifacts)
[Github Actions only triggers on default branch](https://github.com/actions/checkout/discussions/277)
[add-issue-labels-to-pr.yml](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/add-issue-labels-to-pr.yml)

**Comment by Aveline-art** [comment]
* Progress: Currently just opened and started. Have found workflow_run as a safer alternative to pull_request_target. Currently investigating it's Pros and Cons and feasibility for our purposes.
* Blockers: None so far, but once work is done, I would require a second pair of eyes to help with condensing, refactoring, and clarifying the code for maintainability purposes.
* Availability: A bit of time on Sunday, Wednesday, Thursday, and Friday
* ETA: Unsure. Currently performing a lot of feasibility testing.

**Comment by Aveline-art** [comment]
Thoughts on action item: Have the code exit successfully, but quietly in the event of no linked issue

This had been an unexpected development. Having an explicit failure in the event of no linked issue (or incorrectly formatted issue) alerts the creator of the pull request on problems with their pull request. This pushes them to fix the issue rather than wait for a reviewer to explicitly correct them. It might be more valuable, in terms of maintaining workflow, to have it exit loudly and return an automated response.

**Comment by akibrhast** [comment]
>  This pushes them to fix the issue rather than wait for a reviewer to explicitly correct them. It might be more valuable, in terms of maintaining workflow, to have it exit loudly and return an automated response.

Good Point. Given that that's the case, and we actually saw it being helpful in correcting mariannas PR. I am thinking we should leave the failing loudly option that you have currently implemented.

Because the other use case of a pull request not having an issue linked is very rare, and most of the time is done due to some manual and explicit reasons.

**Comment by Aveline-art** [comment]
Notes on action item: Investigate why this action does not run on pull request merging forks to non-default branches (aka wins-feature-1)

Based on testing, I have come to the conclusion that there are two reasons why actions does not trigger on non-default branches.

1) GitHub actions runs the yml files in ```.github/.workflows``` **for the _"into"_ branch**. This means that when the structure of the pull request is "into ```hackforla:wins-feature-1``` from ```aveline-art:adds-demeter-file-to-wins```", the workflow that runs is not the workflow in gh-pages, but wins-feature-1.

2)  The workflow syntax for the ```pull_request``` action requires a ```pull_request.branches``` configuration ([see documentation](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags)). I am still unsure what the exact configuration has to be to trigger a ```pull_request```. It could involve the gitref of the "into" branch, or the "from" branch. Regardless, a branch using wildcards ([also see documentation](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet)) will activate the github action.

As these results are based on testing in my own private repo, if someone else can also test this and leave a comment confirming or rejecting my test results, it would be awesome!

Next steps:

- [x] Discover the exact configuration required for ```pull_request.branches``` so that we may limit use of wildcards
- [ ] Find documentation for point 1 so that this behavior can be officially confirmed

**Comment by Aveline-art** [comment]
Note on future expansion: Although currently not required, should we pursue more complex workflows, the workflow needs to be more modular, especially when jobs or parts of jobs can be reused. This [documentation on separating actions from workflow](https://docs.github.com/en/actions/learn-github-actions/finding-and-customizing-actions#referencing-an-action-in-the-same-repository-where-a-workflow-file-uses-the-action) will be useful.

This will also influence design decisions on file structure. Should file be event-based (this file runs all jobs involving submitted pull requests) or purpose-based (this file runs code needed to add labels to pr). [Link to documentation on managing workflows](https://docs.github.com/en/actions/learn-github-actions/managing-complex-workflows).

**Comment by akibrhast** [comment]
---

> Look for an alternate to the pull_request_target event to prevent malicious code injections into GitHub action execution

I am a bit unsure about the reasoning behind the need to look for an alternative to pull_request_target. Based on [this](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/).
 Which states: 

![image](https://user-images.githubusercontent.com/32349001/123291670-504e3e80-d4e0-11eb-933e-7ceb86bf7d3d.png)


- As you can see, the usage of two workflows and passing around workflow artifacts introduces some overhead. If your workflow scenario simply requires commenting on the PR, but does not require a check out of the modified code, using pull_request_target is a logical shortcut.

- Generally speaking, when the PR contents are treated as passive data, i.e. not in a position of influence over the build/testing process, it is safe. But the repository owners must be extra careful not to trigger any script that may operate on PR controlled contents like in the case of npm install.

Based on the above image and those pointers from the linked article, my comprehension is that while `pull_request_target` has certain drawbacks. It was largely  brought into existence to address the use case of labeling PR's and commenting on PR's and can be considered to be safe as long as its not building and/or processing arbitrary code .

---

> As these results are based on testing in my own private repo, if someone else can also test this and leave a comment confirming or rejecting my test results, it would be awesome!

It would be more convenient for me to test this if I had a copy of your action script. Since you are testing on a private repository. Do you think it would be possible for you to paste the script in a code block in the comments ?

**Comment by Aveline-art** [comment]
@akibrhast I can have my script ready by tomorrow, but it is quite heavy. Since I only tested for triggers, it might be easier to perform the test with [a "hello world" script](https://gist.github.com/Aveline-art/efc0dcdab28ffd46cbe479734ab41624).

**Comment by akibrhast** [comment]
Thanks @Aveline-art , that works. 

Given that the script is getting quite heavy. Have you considered moving away from inline logic to having the code and logic in its own file ?

Something along the lines of this one : https://github.com/hackforla/website/blob/gh-pages/.github/workflows/github-data.yml
Where the action makes use of this script:  https://github.com/hackforla/website/blob/gh-pages/github-actions/github-data/get-project-data.js

**Comment by Aveline-art** [comment]
> Given that the script is getting quite heavy. Have you considered moving away from inline logic to having the code and logic in its own file ?

I have dabbled with making the code more modular, actually, rather than a separate script entirely. I am not sure if a separate script is the right way to go? I would need some guidance on what the best practices would be for the repo. I will make a pull request sometime this weekend to start a discussion on this.

> I am a bit unsure about the reasoning behind the need to look for an alternative to pull_request_target.

Two reasons:

1) It is easier to account for when the script changes. Having to worry about whether something malicious is happening with the script is tiring for the person who codes and the person who reviews. Better safe than sorry.
2) This is more of a personal preference for me, but if a suitable alternative exists, it is worth looking into and investigating so that options can be kept open.

That said, I am unsure if the alternative I have found is actually better, rather than just a security risk in a different form. I would also need guidance on this.

One more note: I think the articles and links we have been using should be added to the wiki, somewhere... They are very critical to the proper functioning and safety of GitHub actions.
